File: solutions/haskell/168-squares-of-a-sorted-array.hs
{-
  Problem 168: Squares of a Sorted Array (LeetCode 977)
  Difficulty: Easy
  Language: Haskell
-}
sortedSquares :: [Int] -> [Int]
sortedSquares nums =
  let arr = listArray (0, length nums - 1) nums
      merge l r
        | l > r = []
        | abs (arr ! l) >= abs (arr ! r) =
            (arr ! l)^2 : merge (l+1) r
        | otherwise = (arr ! r)^2 : merge l (r-1)
  in reverse $ merge 0 (length nums - 1)

-- uses Data.Array

File: solutions/ocaml/168-squares-of-a-sorted-array.ml
(*
  Problem 168: Squares of a Sorted Array (LeetCode 977)
  Difficulty: Easy
  Language: OCaml
*)
let sorted_squares nums =
  let arr = Array.of_list nums in
  let n = Array.length arr in
  let result = Array.make n 0 in
  let rec merge l r i =
    if i < 0 then Array.to_list result
    else if abs arr.(l) >= abs arr.(r) then begin
      result.(i) <- arr.(l) * arr.(l); merge (l + 1) r (i - 1)
    end else begin
      result.(i) <- arr.(r) * arr.(r); merge l (r - 1) (i - 1)
    end
  in merge 0 (n - 1) (n - 1)
